File: hermes-plugin/schemas.py
from __future__ import annotations


ANDROID_SCHEMA = {
    "name": "android",
    "description": (
        "Control Android apps through the local Clawdroid Waydroid daemon. "
        "Use task_route for branded service requests, snapshot plus act for "
        "normal UI interaction, and direct open/url/settings actions when the "
        "destination is already known. Re-snapshot after meaningful UI changes."
    ),
    "parameters": {
        "type": "object",
        "properties": {
            "action": {
                "type": "string",
                "enum": [
                    "status",
                    "current_app",
                    "apps_list",
                    "apps_search",
                    "service_resolve",
                    "task_route",
                    "app_installed",
                    "store_search",
                    "app_open",
                    "activity_start",
                    "intent_start",
                    "url_open",
                    "settings_open",
                    "app_details_open",
                    "market_open",
                    "snapshot",
                    "screenshot",
                    "decide_next",
                    "act",
                    "coordinate_act",
                    "wait",
                ],
            },
            "device": {"type": "string"},
            "store": {"type": "string"},
            "package": {"type": "string"},
            "activity": {"type": "string"},
            "query": {"type": "string"},
            "url": {"type": "string"},
            "intent_action": {"type": "string"},
            "data_url": {"type": "string"},
            "mime_type": {"type": "string"},
            "categories": {"type": "array", "items": {"type": "string"}},
            "extras": {"type": "object", "additionalProperties": {"type": "string"}},
            "settings_action": {"type": "string"},
            "stop": {"type": "boolean"},
            "snapshot_mode": {"type": "string", "enum": ["interactive", "full", "hybrid"]},
            "include_screenshot": {"type": "boolean"},
            "goal": {"type": "string"},
            "decision_mode": {
                "type": "string",
                "enum": ["deterministic", "auto", "llm_text", "llm_vision"],
            },
            "auto_execute": {"type": "boolean"},
            "llm_provider": {"type": "string"},
            "llm_model": {"type": "string"},
            "snapshot_id": {"type": "string"},
            "ref": {"type": "string"},
            "op": {
                "type": "string",
                "enum": [
                    "click",
                    "click_center",
                    "long_click",
                    "set_text",
                    "clear_text",
                    "scroll_forward",
                    "scroll_backward",
                    "scroll_to_start",
                    "scroll_to_end",
                    "press_back",
                    "press_home",
                    "press_recents",
                    "press_enter",
                    "tap",
                    "long_press",
                    "swipe",
                    "type_text",
                ],
            },
            "text": {"type": "string"},
            "x": {"type": "integer"},
            "y": {"type": "integer"},
            "x1": {"type": "integer"},
            "y1": {"type": "integer"},
            "x2": {"type": "integer"},
            "y2": {"type": "integer"},
            "duration_ms": {"type": "integer", "minimum": 0},
            "approved": {"type": "boolean"},
            "limit": {"type": "integer", "minimum": 1, "maximum": 25},
            "timeout_ms": {"type": "integer", "minimum": 0},
            "wait_for": {
                "type": "string",
                "enum": ["idle", "package", "activity", "text", "ref_appears", "ref_gone"],
            },
            "wait_value": {"type": "string"},
        },
        "required": ["action"],
    },
}


ANDROID_ADMIN_SCHEMA = {
    "name": "android_admin",
    "description": (
        "Administrative Clawdroid/Waydroid actions for Hermes. Use only when "
        "the user explicitly asks for host or Android package changes. Direct "
        "APK, store install, remove, extras, and bridge allowlist actions require "
        "approved=true."
    ),
    "parameters": {
        "type": "object",
        "properties": {
            "action": {
                "type": "string",
                "enum": [
                    "doctor",
                    "recover",
                    "waydroid_start",
                    "waydroid_stop",
                    "app_install",
                    "app_install_url",
                    "store_install",
                    "app_remove",
                    "default_stores_install",
                    "device_profile_apply",
                    "extras_install",
                    "extras_uninstall",
                    "bridge_configure",
                ],
            },
            "mode": {"type": "string", "enum": ["user", "system"]},
            "store": {"type": "string"},
            "package": {"type": "string"},
            "query": {"type": "string"},
            "apk_path": {"type": "string"},
            "apk_url": {"type": "string"},
            "extras": {"type": "array", "items": {"type": "string"}},
            "stores": {"type": "array", "items": {"type": "string"}},
            "allowed_packages": {"type": "array", "items": {"type": "string"}},
            "approved": {"type": "boolean"},
            "limit": {"type": "integer", "minimum": 1, "maximum": 25},
            "timeout_ms": {"type": "integer", "minimum": 0},
        },
        "required": ["action"],
    },
}

File: openclaw-plugin/src/schemas.js
import { Type } from "@sinclair/typebox";

const literalUnion = values => Type.Union(values.map(value => Type.Literal(value)));

const ANDROID_ACTIONS = [
  "status",
  "current_app",
  "apps_list",
  "apps_search",
  "service_resolve",
  "task_route",
  "app_installed",
  "store_search",
  "app_open",
  "activity_start",
  "intent_start",
  "url_open",
  "settings_open",
  "app_details_open",
  "market_open",
  "snapshot",
  "screenshot",
  "decide_next",
  "act",
  "coordinate_act",
  "wait"
];

const ANDROID_ADMIN_ACTIONS = [
  "doctor",
  "recover",
  "waydroid_start",
  "waydroid_stop",
  "app_install",
  "app_install_url",
  "store_install",
  "app_remove",
  "default_stores_install",
  "device_profile_apply",
  "extras_install",
  "extras_uninstall",
  "bridge_configure"
];

const SNAPSHOT_MODES = ["interactive", "full", "hybrid"];
const DECISION_MODES = ["deterministic", "auto", "llm_text", "llm_vision"];
const REF_OPS = [
  "click",
  "click_center",
  "long_click",
  "set_text",
  "clear_text",
  "scroll_forward",
  "scroll_backward",
  "scroll_to_start",
  "scroll_to_end",
  "press_back",
  "press_home",
  "press_recents",
  "press_enter",
  "tap",
  "long_press",
  "swipe",
  "type_text"
];
const WAIT_TARGETS = ["idle", "package", "activity", "text", "ref_appears", "ref_gone"];

export const AndroidParameters = Type.Object({
  action: literalUnion(ANDROID_ACTIONS),
  device: Type.Optional(Type.String()),
  store: Type.Optional(Type.String()),
  package: Type.Optional(Type.String()),
  activity: Type.Optional(Type.String()),
  query: Type.Optional(Type.String()),
  url: Type.Optional(Type.String()),
  intent_action: Type.Optional(Type.String()),
  data_url: Type.Optional(Type.String()),
  mime_type: Type.Optional(Type.String()),
  categories: Type.Optional(Type.Array(Type.String())),
  extras: Type.Optional(Type.Record(Type.String(), Type.String())),
  settings_action: Type.Optional(Type.String()),
  stop: Type.Optional(Type.Boolean()),
  snapshot_mode: Type.Optional(literalUnion(SNAPSHOT_MODES)),
  include_screenshot: Type.Optional(Type.Boolean()),
  goal: Type.Optional(Type.String()),
  decision_mode: Type.Optional(literalUnion(DECISION_MODES)),
  auto_execute: Type.Optional(Type.Boolean()),
  llm_provider: Type.Optional(Type.String()),
  llm_model: Type.Optional(Type.String()),
  snapshot_id: Type.Optional(Type.String()),
  ref: Type.Optional(Type.String()),
  op: Type.Optional(literalUnion(REF_OPS)),
  text: Type.Optional(Type.String()),
  x: Type.Optional(Type.Integer()),
  y: Type.Optional(Type.Integer()),
  x1: Type.Optional(Type.Integer()),
  y1: Type.Optional(Type.Integer()),
  x2: Type.Optional(Type.Integer()),
  y2: Type.Optional(Type.Integer()),
  duration_ms: Type.Optional(Type.Integer({ minimum: 0 })),
  approved: Type.Optional(Type.Boolean()),
  limit: Type.Optional(Type.Integer({ minimum: 1, maximum: 25 })),
  timeout_ms: Type.Optional(Type.Integer({ minimum: 0 })),
  wait_for: Type.Optional(literalUnion(WAIT_TARGETS)),
  wait_value: Type.Optional(Type.String())
});

export const AndroidAdminParameters = Type.Object({
  action: literalUnion(ANDROID_ADMIN_ACTIONS),
  mode: Type.Optional(literalUnion(["user", "system"])),
  store: Type.Optional(Type.String()),
  package: Type.Optional(Type.String()),
  query: Type.Optional(Type.String()),
  apk_path: Type.Optional(Type.String()),
  apk_url: Type.Optional(Type.String()),
  extras: Type.Optional(Type.Array(Type.String())),
  stores: Type.Optional(Type.Array(Type.String())),
  allowed_packages: Type.Optional(Type.Array(Type.String())),
  approved: Type.Optional(Type.Boolean()),
  limit: Type.Optional(Type.Integer({ minimum: 1, maximum: 25 })),
  timeout_ms: Type.Optional(Type.Integer({ minimum: 0 }))
});
